drivers: uart: samd0: add missing .configure API functionality

KubaFYI opened the pull request:
The UART driver for samd0 was is missing the .configure
functionality expected from api for serial driver. This commit
fixes this by providing basic configuration.

Note, using the function will briefly switch the peripheral off
taking the data line low.

Signed-off-by: Kuba Sanak <contact@kuba.fyi>

stephanosio commented on drivers/serial/uart_sam0.c:
```diff
@@ -377,6 +377,90 @@ static void uart_sam0_rx_timeout(struct k_work *work)
+	switch (new_cfg->parity) {
+	case UART_CFG_PARITY_NONE:
+		CTRLA_temp.bit.FORM &= 0xe;
+	case UART_CFG_PARITY_ODD:
```
It looks like we are missing `break` here.

stephanosio commented on drivers/serial/uart_sam0.c:
```diff
@@ -377,6 +377,90 @@ static void uart_sam0_rx_timeout(struct k_work *work)
 #endif
 
+static int uart_sam0_configure(struct device *dev,
+								const struct uart_config *new_cfg)
```
could probably use less TABs

KubaFYI replied:
While it looks good on my text editor I agree that it might be a bit much for ones with more than 4 spaces per tab. Will change.

stephanosio commented on drivers/serial/uart_sam0.c:
```diff
@@ -377,6 +377,90 @@ static void uart_sam0_rx_timeout(struct k_work *work)
+{
+	int retval;
+	const struct uart_sam0_dev_cfg *const cfg = DEV_CFG(dev);
+	SercomUsart * const usart = cfg->regs;
```
We should probably stick with either `*const` or `* const`.

KubaFYI replied:
I agree. I initially went with `*const` but that ended with checkpatch shouting at me and not passing the test, so I changed it to `* const`. Unfortunately the whole file has inconsistencies in how `SercomUsart *` declarations are formatted, which could be fixed but I wanted to avoid touching lines not related to the functionality I implemented.

stephanosio replied:
Alright, it looks like we will need to fix checkpatch first then.

Marking it as resolved (won't fix).

stephanosio commented on drivers/serial/uart_sam0.c:
```diff
@@ -377,6 +377,90 @@ static void uart_sam0_rx_timeout(struct k_work *work)
+
+	switch (new_cfg->parity) {
+	case UART_CFG_PARITY_NONE:
+		CTRLA_temp.bit.FORM &= 0xe;
```
Since we do not support "auto-baud", we should just set `FORM` to 0; in case, the SERCOM was previously configured (e.g. by bootloader).

KubaFYI replied:
I've mostly wanted to preserve any config the user might have done previously. Though, I thought any leftover bootloader settings would be overwritten [in the init function](https://github.com/zephyrproject-rtos/zephyr/blob/c25357a24764873760622ef96cfe678907d3f3e1/drivers/serial/uart_sam0.c#L491) anyway?

stephanosio replied:
> bootloader settings would be overwritten [in the init function](https://github.com/zephyrproject-rtos/zephyr/blob/c25357a24764873760622ef96cfe678907d3f3e1/drivers/serial/uart_sam0.c#L491) anyway?

You are right. Bootloader setting will not be a problem then.

I see another issue with this approach though, that LSB being set does not necessarily mean parity being enabled (although this pattern does seem to apply to the SAM D2x, even though there is no explicit mention of such in the datasheet).

For instance, refer to the `FORM` values for SAM D5x below:
![image](https://user-images.githubusercontent.com/11345418/76089432-fed52400-5ffc-11ea-9552-c0d1d868cdd8.png)

I wonder if we should handle this in a more verbose way?

KubaFYI replied:
Good catch, I've only based this change on the samd2x datasheet. Even though the ISO7816 mode is not really used in the driver (and neither is the auto-baud one) it might indeed make sense to handle this in a more verbose way for the sake of future proofing it.

stephanosio commented on drivers/serial/uart_sam0.c:
```diff
@@ -377,6 +377,90 @@ static void uart_sam0_rx_timeout(struct k_work *work)
+	case UART_CFG_PARITY_NONE:
+		CTRLA_temp.bit.FORM &= 0xe;
+	case UART_CFG_PARITY_ODD:
+		CTRLA_temp.bit.FORM |= 0x1;
```
same here, simply set `FORM` to 1.

stephanosio commented on drivers/serial/uart_sam0.c:
```diff
@@ -377,6 +377,90 @@ static void uart_sam0_rx_timeout(struct k_work *work)
+		CTRLB_temp.bit.PMODE = 1;
+		break;
+	case UART_CFG_PARITY_EVEN:
+		CTRLA_temp.bit.FORM |= 0x1;
```
ditto

stephanosio commented on drivers/serial/uart_sam0.c:
```diff
@@ -377,6 +377,93 @@ static void uart_sam0_rx_timeout(struct k_work *work)
 #endif
 
+static int uart_sam0_configure(struct device *dev,
+					const struct uart_config *new_cfg)
```
Nit: Indentation

To be consistent with the rest of the file, we should have:
```
static int uart_sam0_configure(struct device *dev,
			       const struct uart_config *new_cfg)
```

Note that the tab width is 8.

stephanosio reviewed (commented):
This PR adds the `.configure` interface implementation without `.config_get`. Both should be implemented to pass the UART driver test.

p.s. As far as I can see, all the other UART drivers that support `.configure` implement `.config_get` as well.

nandojve commented:
We've been converting to new DT macros all Atmel SAM variations. The #24110 is about SAM Serial Drivers. I can rebase on top of it and add changes. I waiting some feedback on #24110 anyway but is important know if this can be merged soon.

stephanosio reviewed (commented):
The conventional way of implementing `config_get` would be to cache configurations in `struct uart_config` or data variables in the driver context and simply return those values.

The assumption is, and should be, that an application will not make arbitrary direct configuration changes to the UART hardware.

(e.g. see `uart_ns16550.c`, `uart_nrfx_uart.c` and others)

p.s. ~I am neither +1 nor -1 on the current implementation~; the API specification does not state exactly how it should be implemented. With that being the case, the current implementation would cause a problem if an application relies on the behaviour in which the configured and retrieved baudrates exactly match, or the retrieved baudrates are expected to match one of the preset values.

EDIT:
> the API specification does not state exactly how it should be implemented

though the test (`tests/drivers/uart/uart_basic_api`) does.

stephanosio reviewed (changes requested):
As far as the UART driver test (`tests/drivers/uart/uart_basic_api`) goes, the configured and retrieved data MUST EXACTLY MATCH:
https://github.com/zephyrproject-rtos/zephyr/blob/608dba486eebef39317b4e680be36abef297a0cd/tests/drivers/uart/uart_basic_api/src/test_uart_config.c#L79

So the current implementation will not pass the test.

jenmwms commented on drivers/serial/uart_sam0.c:
```diff
@@ -90,6 +90,29 @@ static void wait_synchronization(SercomUsart *const usart)
+
+	baud = usart->BAUD.reg;
+
+	tmp = 65536 - baud;
```
Why is this hard-coded at 65536 here? Would it be better to define a local variable with this value? A brief description would help at least.

KubaFYI replied:
I followed the lead of the already defined `uart_sam0_set_baudrate` function. I was considering turning it into a variable but then I'd need to touch the set_baudrate function as well even though my changes are not related to it. What's the usual approach in a situation like this? I'm assuming that in-file consistency matters (so both get_baudrate and set_baudrate should do the same thing with the magic number) - in that case should one stick with the already-present, ugly, hard-coded number or define a variable and change the rest of the file which wouldn't normally be touched given the functionality change?

KubaFYI commented:
@stephanosio Ah, this branch was actually created before the `test_uart_config.c` file even existed so I didn't know the `uart_config`s are expected to match exactly. Will fix that now.

stephanosio commented on drivers/serial/uart_sam0.c:
```diff
@@ -71,6 +72,8 @@ struct uart_sam0_dev_data {
 #endif
 };
 
+static struct uart_config config_cache;
```
This is incorrect -- it cannot support multiple UART device instances.

The `config_cache` should be part of the device context (i.e. `struct uart_sam0_dev_cfg` or `struct uart_sam0_dev_data`) to store the config values for each UART device instance.

KubaFYI replied:
Done.

KubaFYI committed:
drivers: uart: samd0: add missing .configure API functionality

The UART driver for samd0 was is missing the .configure and
.config_get functionality expected from api for serial driver.
This commit fixes this by providing basic configuration

Signed-off-by: Kuba Sanak <contact@kuba.fyi>

stephanosio reviewed (approved):
Looks ok to me.

jenmwms commented on drivers/serial/uart_sam0.c:
```diff
@@ -377,10 +380,121 @@ static void uart_sam0_rx_timeout(struct k_work *work)
+	return 0;
+}
+
+static int uart_sam0_config_get(struct device *dev,
```
jw - Is it worth considering a way to accomplish this task without using `memcpy()`? I ask because I am curious about it from a secure coding perspective atm. Can this use of memcpy introduce a vulnerability? Please see uart_nrfx_uart.c and uart_ns16550.c for example implementations.

KubaFYI replied:
In choosing the `memcpy()` I was aiming for making absolutely sure that the implementation passes relevant tests which currently use `memcmp()` to establish whether the set and retrieved `struct uart_config`s match. Doing things the way uart_nrfx_uart.c or uart_ns16550.c do it can potentially introduce a corner case where the test won't pass because of potential content of padding in the struct which won't be carried across without `memcpy()`.

I guess the proper thing to do would be to edit the tests to use comparison of individual struct member variables instead of `memcmp()`, but I wanted to avoid doing changes not related to the work I'm doing.

jenmwms replied:
Ah that makes sense, thanks for clarifying! I am learning more with your help.
I wonder if you might want to add the cast to void in anticipation of the MISRA-C coding guidelines currently in discussion, namely regarding memcpy --> `(void) memcpy()` (see #18344). Adding @ceolin in case he wants to share thoughts here

stephanosio replied:
As long as the copy size is static (i.e. `sizeof(dev_data->config_cache)`), that is highly unlikely.

As for the `out_cfg` pointer checking, that is a whole another issue.

stephanosio replied:
Enforcement of MISRA-C, especially arguably ridiculous ones like `(void)func()` is a no-go for now IIRC.

ceolin replied:
Hi @jenmwms, regarding MISRA, it is not sepcific to `memcpy`. The rule says that we should check all return values from non-void functions. It happens, that memcpy() return is quite useless, so we just explicitly ignore it casting to void.

KubaFYI replied:
@stephanosio

> As for the `out_cfg` pointer checking, that is a whole another issue.

Do you mean the checking in function or in the test. In short, is there any other changes I should make to this PR?

stephanosio replied:
@KubaFYI As far as my comments are concerned, no changes necessary in this PR.

I was simply alluding that we can go very far with advanced security checks and functional safety rules; but, given that the current Zephyr codebase is really not ready for compliance, I see no reason to enforce them in this particular PR.
